sklearn/linear_model/tests/test_quantile.py

```diff
@@ -0,0 +1,126 @@
+                           noise=1.0)
+    for q in [0.5, 0.9, 0.05]:
+        quant = QuantileRegressor(quantile=q, alpha=0).fit(X, y)
+        fraction_below = np.mean(y < quant.predict(X))
+        assert_allclose(fraction_below, q, atol=1e-2)
+
+
+@pytest.mark.skipif(sp_version < parse_version("1.0.0"), reason=_SCIPY_TOO_OLD)
+def test_quantile_without_intercept():
+    X, y = make_regression(n_samples=300, n_features=20, random_state=0,
+                           noise=1.0)
+    quant = QuantileRegressor(alpha=1e-4, fit_intercept=False).fit(X, y)
+    # check that result is similar to Huber
+    huber = HuberRegressor(epsilon=1 + 1e-4, alpha=1e-4, fit_intercept=False
+                           ).fit(X, y)
+    assert_allclose(huber.intercept_, quant.intercept_, atol=1e-1)
+    assert_allclose(huber.coef_, quant.coef_, atol=1e-1)
+    # check that we still predict fraction
+    fraction_below = np.mean(y < quant.predict(X))
+    assert_allclose(fraction_below, 0.5, atol=1e-1)
+
+
+@pytest.mark.skipif(sp_version < parse_version("1.0.0"), reason=_SCIPY_TOO_OLD)
+def test_quantile_sample_weight():
+    # test that with unequal sample weights we still estimate weighted fraction
+    n = 1000
+    X, y = make_regression(n_samples=n, n_features=10, random_state=0,
+                           noise=10.0)
+    weight = np.ones(n)
+    # when we increase weight of upper observaions,
+    # estimate of quantile should go up
+    weight[y > y.mean()] = 100
+    quant = QuantileRegressor(quantile=0.5, alpha=1e-4)
+    quant.fit(X, y, sample_weight=weight)
+    fraction_below = np.mean(y < quant.predict(X))
+    assert fraction_below > 0.5
+    weighted_fraction_below = np.sum((y < quant.predict(X)) * weight) \
+        / np.sum(weight)
+    assert_allclose(weighted_fraction_below, 0.5, atol=1e-2)
+
+
+@pytest.mark.skipif(sp_version < parse_version("1.0.0"), reason=_SCIPY_TOO_OLD)
+@pytest.mark.parametrize('quantile', [2.0, 1.0, 0.0, -1])
+def test_quantile_incorrect_quantile(quantile):
+    X, y = make_regression(n_samples=10, n_features=1, random_state=0, noise=1)
+    with assert_raises(ValueError):
+        QuantileRegressor(quantile=quantile).fit(X, y)
+
+
+@pytest.mark.skipif(sp_version < parse_version("1.0.0"), reason=_SCIPY_TOO_OLD)
+@pytest.mark.parametrize('quantile', [0.1, 0.5, 0.9])
+def test_asymmetric_error(quantile):
+    n_samples = 1000
+    n_features = 3
+    bias = 12.3
+    param = 1.0
+    generator = check_random_state(42)
+    X = generator.randn(n_samples, n_features)
+    ground_truth = generator.rand(n_features)
+    dist = scipy.stats.expon(param)
+    noise = dist.rvs(size=n_samples, random_state=42)
+    y = np.dot(X, ground_truth) + bias + noise
+    model = QuantileRegressor(quantile=quantile).fit(X, y)
+
+    assert_allclose(model.intercept_, bias + dist.ppf(quantile), atol=0.1)
+    assert_allclose(model.coef_, ground_truth, atol=0.3)
+    assert_allclose(np.mean(model.predict(X) > y), quantile, atol=0.003)
```

Comment:
```suggestion
    X = np.abs(rng.randn(n_samples, n_features))  # make it positive
    coef = np.abs(rng.rand(n_features))  # make it positive
    # For an exponential distr with rate lambda, e.g. exp(-lambda*x), the quantile at level q is:
    # quantile(q) = - log(1 - q) / lambda
    # scale = 1/lambda = -quantile(q) / log(1-q)
    y = rng.exponential(scale = -(dot(X, coef) + intercept) / np.log(1 - quantile), size=n_samples)
    model = QuantileRegressor(quantile=quantile, tol=...).fit(X, y)

    assert_allclose(model.intercept_, intercept)
    assert_allclose(model.coef_, coef)
    assert_allclose(np.mean(model.predict(X) > y), quantile, atol=1/n_samples)
```
This way, we have `quantile = dot(X, coef) + intercept`. We have to take care, that this is a positive quantity.